### DISCUSSION TITLE
I Used Tasmota / Thermostat to replace fridge thermostat

### DISCUSSION THREAD

**Discussion by bibikiwi** [arendst/Tasmota#12386] [Show and tell]
Lately it has been very hot and I noticed that my decade old beverages fridge / wine cooler started having trouble keeping its insides cool and a few days ago it all became too much for it and it died. I opened it up, checked the circuitry and it turned out to be the thermostat module (a purely mechanical one). 
Since in my house I have already tasmotized everything else there is to tasmotize, I decided to give the Thermostat variant a try.

see : https://tasmota.github.io/docs/Thermostat/
it says "testers are welcomed", so here is my review : 

I'm very impressed by that "intelligent control strategy" as apart from switching it to cooling mode (CLIMATEMODESET 1) I only had to set the target temperature (TEMPTARGETSET 15) and see it work. It keeps its temperature of around 15°C in a room with about 28°C with an average duty cycle of about 30%. And it consumes much less now ! Plus the advantage of being able to monitor and control it in home assistant, grafana, etc...

On the doc page, they explain how to make the parameters persistent through reboots and I can confirm that works too.
I could have imagined having a power cut, wifi not coming up and the fridge going haywire because not being able to fetch parameters, but I can confirm that it has passed that test beautifully too.

It can fetch its parameters from persistent MQTT  topics and that's pretty neat if you want to control that e.g. from a widget in home assistant.

I used an ESP8266 (a wemos D1), a relay hat and a waterproofed variant of the DS18b20 temperature probe.

If you try this yourself, watch out not to electrocute yourself !

![Screenshot from 2021-06-16 21-17-05](https://user-images.githubusercontent.com/32168005/122283721-21870580-ceed-11eb-8f33-829304aa36a8.png)
![WhatsApp Image 2021-06-12 at 12 08 08](https://user-images.githubusercontent.com/32168005/122283785-306db800-ceed-11eb-8437-e7e2885e0a3a.jpeg)
![WhatsApp Image 2021-06-12 at 12 09 19 (1)](https://user-images.githubusercontent.com/32168005/122283813-395e8980-ceed-11eb-9b9a-96b1b8ae56a5.jpeg)
![WhatsApp Image 2021-06-12 at 12 09 34](https://user-images.githubusercontent.com/32168005/122283823-3cf21080-ceed-11eb-8b83-a7567ab17801.jpeg)
![WhatsApp Image 2021-06-12 at 12 09 49](https://user-images.githubusercontent.com/32168005/122283832-3fed0100-ceed-11eb-8a0b-35a93372c366.jpeg)
![WhatsApp Image 2021-06-12 at 12 10 27](https://user-images.githubusercontent.com/32168005/122283849-47140f00-ceed-11eb-8d83-97a08fc1dc91.jpeg)
![WhatsApp Image 2021-06-12 at 12 08 08 (1)](https://user-images.githubusercontent.com/32168005/122283874-5004e080-ceed-11eb-80e1-e5baa116d1ec.jpeg)

**Comment by sfromis** [comment]
Nice example of getting something to work, and improve it at the same time.

**Reply by barbudor** [reply]
Nicely done @bibikiwi  👏

**Reply by blakadder** [reply]
Very nice indeed!

**Reply by Jason2866** [reply]
Wine needs the right temperature. So this project had top priority.
Nice job done!

**Reply by maximr377** [reply]
good afternoon, I congratulate you on the project. I tell you that I am a newbie in home automation. I want to make a thermostat with tasmota for my fridge and I got to the page and I don't know where to start, will you mind being able to share your .bin compilation? my idea is to work it and add a display because I want to install it on an esp32 with oled

**Reply by bibikiwi** [reply]
@maximr377 
this is what I did : 
I read https://tasmota.github.io/docs/Thermostat/

download the tasmota repo

    git clone https://github.com/arendst/Tasmota.git

go inside the new folder and enable the user config override

    cd Tasmota
    cp tasmota/user_config_override_sample.h  tasmota/user_config_override.h
    
edit the new file and add this almost at the end (where it says "Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif")

     #ifndef USE_THERMOSTAT
     #define USE_THERMOSTAT
     #endif

now compile and upload (how this works is a bit dependant on what you usually use for making arduino projects, I use vscode with platformio on ubuntu, so the following might need some fiddling if you have another setup)



install platformio (the one from ubuntu repo is not updated frequently enough, and the pip is often a headbreaker, So if you have one of those already installed and it doesn't work well, remove them first, and reinstall via the script. I found that the easiest way to install it is the script from the platformio docs ( https://docs.platformio.org/en/latest/core/installation.html )

     python3 -c "$(curl -fsSL https://raw.githubusercontent.com/platformio/platformio/master/scripts/get-platformio.py)"

Connect your arduino (wemos D1 for me) and look at what tty it connects to (dmesg)
finetune your setup (wemos D1 board, usb port /dev/ttyUSB0, etc...) edit following file based on the included sample

     cp platformio_override_sample.ini platformio_override.ini
     
I only changed : 

      upload_port                 = /dev/ttyUSB0

and at the end added 

     [core]
     platform = espressif8266

then  I just run

     platformio run --target upload

and I sit back and relax while I watch it compile and upload to flash.

When it is finished, it acts like a normal Tasmota. You connect to its wifi access point with your cellphone, fill in the wifi creds of your home wifi and then it will connect and give you the new ip address. Connect your cellphone back to your home wifi and you can then surf to the tasmota interface.
Then execute the following steps on the Tasmota console (easier to do that on a PC instead of your cellphone)

enable the thermostat driver

     THERMOSTATMODESET 1

by default the thermostat algorithm is set to heating. Switch it to cooling 

     CLIMATEMODESET 1

**Reply by erniberni** [reply]
Hi, 
nice project ...
In your Grafana I see a control difference (delta) of approximate 2°C, correct?
Which thermostat parameters differ from the default setup?
I am working on a heater project and my control difference is up to now 4°C - that needs to be improved.

**Reply by bibikiwi** [reply]
@erniberni Sorry, I'm not sure what you mean with "control difference (delta)", but if you mean "the difference between the desired temperature and the actual temperature", I did not choose myself at what temperatures to trigger, as you can see in the grafana, it does not trigger on preset high and low temperatures, but it's rather a "PID" algorithm that tunes the duty cycle depending on the deviation from the desired temperature. It's pretty efficient and for sure way more efficient than the bi-metal switch that was driving it before.
What I changed is exactly described in this post.

**Reply by sfromis** [reply]
The [Thermostat feature](https://tasmota.github.io/docs/Commands/#thermostat) does have a lot of parameters you can set, including some which look like being related to temperature change sensitivity, but I've not worked with it.

**Comment by grappini** [comment]
How often are the temperatures read in home assistant?

**Reply by sfromis** [reply]
In principle never, as the usual setup is for Tasmota to publish MQTT payloads including sensor data, and the backend tool (HA or whatever) to be subscribing to these messages. The frequency of that can be controlled by the TelePeriod command, default every 5 minutes.

**Reply by grappini** [reply]
So the average temperatures calculated in the screnshots above are based on 5 min intervals.

**Reply by bibikiwi** [reply]
I can confirm confirm this, I did not touch the TelePeriod setting. 5 minutes. This is a snippet from the log

     Jul 25 18:51:53 wijnkoeler_17B3AD-5037 ESP-MQT: homeassistant/tele/wijnkoeler_17B3AD/SENSOR = {"Time":"2021-07-25T17:51:52","DS18B20":{"Id":"051693DD7BFF","Temperature":15.9},"TempUnit":"C"}
     Jul 25 18:56:53 wijnkoeler_17B3AD-5037 ESP-MQT: homeassistant/tele/wijnkoeler_17B3AD/SENSOR = {"Time":"2021-07-25T17:56:52","DS18B20":{"Id":"051693DD7BFF","Temperature":16.7},"TempUnit":"C"}
     Jul 25 19:01:53 wijnkoeler_17B3AD-5037 ESP-MQT: homeassistant/tele/wijnkoeler_17B3AD/SENSOR = {"Time":"2021-07-25T18:01:52","DS18B20":{"Id":"051693DD7BFF","Temperature":17.4},"TempUnit":"C"}
     Jul 25 19:06:53 wijnkoeler_17B3AD-5037 ESP-MQT: homeassistant/tele/wijnkoeler_17B3AD/SENSOR = {"Time":"2021-07-25T18:06:52","DS18B20":{"Id":"051693DD7BFF","Temperature":17.9},"TempUnit":"C"}
     Jul 25 19:11:53 wijnkoeler_17B3AD-5037 ESP-MQT: homeassistant/tele/wijnkoeler_17B3AD/SENSOR = {"Time":"2021-07-25T18:11:52","DS18B20":{"Id":"051693DD7BFF","Temperature":17.3},"TempUnit":"C"}

The relay state however, is sent immediately when it changes.

**Comment by grappini** [comment]
Does this thermostat function work without room temperature sensor ? Just with the fridge temp sensor ? Or it requires both sensors ?

**Reply by dagelf** [reply]
Why would it need a room sensor? I can't think of any useful info that a room sensor can feed into it.  On the fun side... you can measure the effect of the fridge radiator on the room... :smile:

**Reply by sfromis** [reply]
Room sensor is probably just because it is nice to have, when it can be hooked up very easily, without being part of the thermostat function. Edge case, you could avoid turning on the fridge if ambient temperature is low enough.

**Reply by dagelf** [reply]
That would probably be a bad idea - besides I love the fact that the "defaults" and the simplest approach worked better than the original thermostat. Question: does the processing take place on the ESP32 itself? (Ie. there's no chance of a server or Wi-Fi being down, causing your fridge to overheat?)

**Reply by sfromis** [reply]
Since the thermostat feature of Tasmota is used, the temperature control is local on the ESP8266. ESP32 could also have been used.

**Reply by bibikiwi** [reply]
I added the room sensor afterwards and have since then even added one of those reed sensors (magnetic switch) too. Indeed, just because why not ? It doesn't influence the thermostat behaviour, but it's interesting to have.

I added them to the graph and so I can see how the ambient temperature influences how hard the fridge has to work or how the temperature rises when I leave it open for a few minutes while restocking it or if someone doesn't close it well, I get an HA notification. The same room also contains other fridges, freezers, washing machine, dryer and water heater and central heating boiler. e.g. when the dryer is active, it's like a sauna in there.

And as I explained in my original post, it was pretty important for me that it would not go crazy if e.g. it couldn't reach the wifi/mqtt/homeassistant, I have tested that and it works as hoped : The docs explain how you can set any settings as defaults on bootup in Tasmota. On the other hand it's also nice to be able to make Tasmota read any setting from MQTT, e.g. so I can change the desired temperature from an HA dashboard (although, how often would one need to do that ?)

**Comment by gemi254** [comment]
I try to use thermostat and set temperature for a deep freezer..
In console when i type..
TEMPTARGETSET -1.5
i get..
stat/tasmota_65FF14/RESULT = {"TempTargetSet1":4.0}

Seems tha minimum allowed temp target for thermostat is 4.0 C
Is a way to set negative temp target?

**Reply by sfromis** [reply]
The minimum allowable temperature can be set via `TempFrostProtectSet`, see [commands](https://tasmota.github.io/docs/Commands/#thermostat).
